File: macos/MacEcho/Storage/TrustStore.swift
// TrustStore.swift — Phase 9
//
// Encrypted local storage for trusted device metadata per
// 04_SECURITY_MODEL.md §Local Trust Database and Architecture Decision 18.
//
// This is NOT the Apple Keychain (that is KeychainManager.swift).
// The Security Model explicitly separates:
//   - Private keys          → Apple Keychain (KeychainManager.swift)
//   - Trust metadata        → Encrypted local storage (this file)
//
// Trust metadata stored per entry (04_SECURITY_MODEL.md §Local Trust Database):
//   • Trusted Device Identifier
//   • Trusted Public Key (X25519 + Ed25519 raw representations)
//   • Pairing Timestamp
//   • Trust Status
//   • Device Metadata (name, type)
//
// Encryption:
//   A dedicated AES-256-GCM symmetric key is stored in the Keychain under the
//   "com.macecho.app.truststore" service. All trust data is encrypted with this
//   key before writing to the Application Support directory. The trust store key
//   is separate from the identity key (Architecture Decision 18: "Separating
//   metadata into encrypted local storage was preferred, since metadata does not
//   require hardware-backed key storage and benefits from simpler access patterns.")
//
// Serialization: Swift Codable → JSONEncoder → AES-GCM encrypted Data → file.
//
// Must NOT contain:
//   - Pairing logic          → Phase 12 (TrustStore is populated by Phase 12)
//   - Authentication logic   → Phase 13
//   - Transport              → Phase 7
//   - Private key material   → KeychainManager.swift
//   - AppKit, NSView, menu bar code
//   - Business logic of any kind
//
// Starts empty. Phase 12 (pairing) will call addOrUpdate(_:) to populate it.
//
// Storage file: ~/Library/Application Support/MacEcho/macecho_trust.enc
// Key in Keychain service: "com.macecho.app.truststore"
// Key account name: "trust.encryption.key"

import CryptoKit
import Foundation
import Security

// ---------------------------------------------------------------------------
// TrustStatus
// ---------------------------------------------------------------------------

/// Current trust status of a paired remote device.
///
/// - `trusted`: Pairing completed; device is recognized as a trusted peer.
/// - `revoked`: Trust removed by unpairing, key change, or explicit revocation.
enum TrustStatus: String, Codable, CaseIterable {
    case trusted = "TRUSTED"
    case revoked = "REVOKED"
}

// ---------------------------------------------------------------------------
// TrustEntry — one record in the local trust database
// ---------------------------------------------------------------------------

/// A single entry in the local trust database.
///
/// Represents one trusted (or previously trusted) remote device. Populated
/// by Phase 12 (pairing) and validated by Phase 13 (authentication).
///
/// Per `04_SECURITY_MODEL.md §Local Trust Database`:
/// - Never stores the other device's private key.
/// - The backend has no authority to modify this record.
/// - Only the local application may update trust records after successful
///   security validation.
struct TrustEntry: Codable, Equatable {
    /// The peer's cryptographically random device ID.
    let trustedDeviceId: String
    /// Peer's X25519 public key raw representation (32 bytes).
    let trustedX25519PublicKeyData: Data
    /// Peer's Ed25519 public key raw representation (32 bytes).
    let trustedEd25519PublicKeyData: Data
    /// Unix epoch milliseconds when pairing completed.
    let pairingTimestampMs: Int64
    /// Current trust status.
    let trustStatus: TrustStatus
    /// Human-readable peer device name (e.g. "Suyesh's Android").
    let deviceName: String
    /// Peer platform string (e.g. "ANDROID").
    let deviceType: String
}

// ---------------------------------------------------------------------------
// TrustStore errors
// ---------------------------------------------------------------------------

/// Errors that can occur during TrustStore operations.
enum TrustStoreError: Error, LocalizedError {
    case keychainKeyWriteFailed(OSStatus)
    case keychainKeyReadFailed
    case encryptionFailed
    case decryptionFailed
    case fileWriteFailed(Error)
    case fileReadFailed(Error)

    var errorDescription: String? {
        switch self {
        case .keychainKeyWriteFailed(let status): return "Trust key Keychain write failed: OSStatus \(status)"
        case .keychainKeyReadFailed:              return "Trust encryption key not found in Keychain"
        case .encryptionFailed:                   return "Trust data encryption failed"
        case .decryptionFailed:                   return "Trust data decryption failed — data may be corrupt"
        case .fileWriteFailed(let e):             return "Trust store file write failed: \(e.localizedDescription)"
        case .fileReadFailed(let e):              return "Trust store file read failed: \(e.localizedDescription)"
        }
    }
}

// ---------------------------------------------------------------------------
// TrustStore
// ---------------------------------------------------------------------------

/// Encrypted local storage for trusted device metadata.
///
/// Operations are synchronous and safe to call from any thread. In production,
/// callers should dispatch to a background queue (e.g. `DispatchQueue.global(qos: .utility)`).
/// Thread safety is the caller's responsibility.
struct TrustStore {

    private static let keychainService = "com.macecho.app.truststore"
    private static let trustKeyAccount = "trust.encryption.key"
    private static let trustFileName   = "macecho_trust.enc"
    private static let appSupportSubdir = "MacEcho"

    // -----------------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------------

    /// Adds or updates a trust entry.
    ///
    /// If an entry with `entry.trustedDeviceId` already exists, it is replaced.
    /// Called by Phase 12 (pairing) after a successful pairing flow.
    ///
    /// - Throws: `TrustStoreError` if encryption or file write fails.
    func addOrUpdate(_ entry: TrustEntry) throws {
        var entries = try loadEntries()
        entries[entry.trustedDeviceId] = entry
        try saveEntries(entries)
    }

    /// Returns the trust entry for `deviceId`, or nil if not found.
    ///
    /// Called by Phase 13 (authentication) to verify an incoming connection.
    func get(deviceId: String) -> TrustEntry? {
        (try? loadEntries())?[deviceId]
    }

    /// Returns all trust entries in the store. Order is not guaranteed.
    /// Returns an empty array if no devices have been paired.
    func getAll() -> [TrustEntry] {
        (try? loadEntries())?.values.map { $0 } ?? []
    }

    /// Returns `true` if a trust entry exists for `deviceId`.
    func contains(deviceId: String) -> Bool {
        (try? loadEntries())?[deviceId] != nil
    }

    /// Removes the trust entry for `deviceId`.
    ///
    /// Safe to call when `deviceId` is not present (idempotent).
    /// Called by Phase 12 (unpairing flow).
    ///
    /// - Throws: `TrustStoreError` if encryption or file write fails.
    func remove(deviceId: String) throws {
        var entries = try loadEntries()
        if entries.removeValue(forKey: deviceId) != nil {
            try saveEntries(entries)
        }
    }

    /// Removes all trust entries. The trust store file is overwritten with
    /// an empty encrypted store. Used by application reset flows.
    ///
    /// - Throws: `TrustStoreError` if file write fails.
    func clear() throws {
        try saveEntries([:])
    }

    /// Returns the number of trust entries currently stored.
    func count() -> Int {
        (try? loadEntries())?.count ?? 0
    }

    // -----------------------------------------------------------------------
    // Private implementation
    // -----------------------------------------------------------------------

    private func loadEntries() throws -> [String: TrustEntry] {
        let fileURL = trustStoreURL()
        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            return [:]
        }
        let encryptedData: Data
        do {
            encryptedData = try Data(contentsOf: fileURL)
        } catch {
            throw TrustStoreError.fileReadFailed(error)
        }
        let key = try getOrCreateTrustKey()
        let plaintext = try decrypt(encryptedData, key: key)
        let entries = try JSONDecoder().decode([String: TrustEntry].self, from: plaintext)
        return entries
    }

    private func saveEntries(_ entries: [String: TrustEntry]) throws {
        let key = try getOrCreateTrustKey()
        let plaintext: Data
        do {
            plaintext = try JSONEncoder().encode(entries)
        } catch {
            throw TrustStoreError.encryptionFailed
        }
        let encryptedData = try encrypt(plaintext, key: key)
        let fileURL = trustStoreURL()
        try ensureParentDirectory(for: fileURL)
        do {
            try encryptedData.write(to: fileURL, options: .atomic)
        } catch {
            throw TrustStoreError.fileWriteFailed(error)
        }
    }

    // -----------------------------------------------------------------------
    // Encryption (CryptoKit AES-256-GCM)
    // -----------------------------------------------------------------------

    private func encrypt(_ plaintext: Data, key: SymmetricKey) throws -> Data {
        do {
            let sealedBox = try AES.GCM.seal(plaintext, using: key)
            // combined: nonce + ciphertext + tag — self-contained for storage
            guard let combined = sealedBox.combined else {
                throw TrustStoreError.encryptionFailed
            }
            return combined
        } catch {
            throw TrustStoreError.encryptionFailed
        }
    }

    private func decrypt(_ data: Data, key: SymmetricKey) throws -> Data {
        do {
            let sealedBox = try AES.GCM.SealedBox(combined: data)
            return try AES.GCM.open(sealedBox, using: key)
        } catch {
            throw TrustStoreError.decryptionFailed
        }
    }

    // -----------------------------------------------------------------------
    // Trust encryption key (Keychain-backed)
    // -----------------------------------------------------------------------

    /// Returns the existing trust encryption key from Keychain, or generates
    /// a new 256-bit AES key and stores it in Keychain.
    ///
    /// This key is separate from the identity private keys in KeychainManager.
    /// Per Architecture Decision 18: "Trust metadata does not require hardware-backed
    /// key storage and benefits from simpler access patterns."
    private func getOrCreateTrustKey() throws -> SymmetricKey {
        // Try reading existing key
        if let keyData = keychainReadKey() {
            return SymmetricKey(data: keyData)
        }
        // Generate a new 256-bit key using CryptoKit (system CSPRNG)
        let newKey = SymmetricKey(size: .bits256)
        let keyData = newKey.withUnsafeBytes { Data($0) }
        try keychainStoreKey(keyData)
        return newKey
    }

    private func keychainStoreKey(_ keyData: Data) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: TrustStore.keychainService,
            kSecAttrAccount as String: TrustStore.trustKeyAccount,
            kSecValueData as String: keyData,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
            kSecAttrSynchronizable as String: kCFBooleanFalse!,
        ]
        let addStatus = SecItemAdd(query as CFDictionary, nil)
        if addStatus == errSecSuccess { return }
        if addStatus == errSecDuplicateItem {
            let searchQuery: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrService as String: TrustStore.keychainService,
                kSecAttrAccount as String: TrustStore.trustKeyAccount,
            ]
            let updateAttr: [String: Any] = [kSecValueData as String: keyData]
            let updateStatus = SecItemUpdate(searchQuery as CFDictionary, updateAttr as CFDictionary)
            if updateStatus == errSecSuccess { return }
            throw TrustStoreError.keychainKeyWriteFailed(updateStatus)
        }
        throw TrustStoreError.keychainKeyWriteFailed(addStatus)
    }

    private func keychainReadKey() -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: TrustStore.keychainService,
            kSecAttrAccount as String: TrustStore.trustKeyAccount,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
            kSecAttrSynchronizable as String: kCFBooleanFalse!,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess else { return nil }
        return result as? Data
    }

    // -----------------------------------------------------------------------
    // File system
    // -----------------------------------------------------------------------

    private func trustStoreURL() -> URL {
        let appSupport = FileManager.default.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first!
        return appSupport
            .appendingPathComponent(TrustStore.appSupportSubdir)
            .appendingPathComponent(TrustStore.trustFileName)
    }

    private func ensureParentDirectory(for url: URL) throws {
        let dir = url.deletingLastPathComponent()
        if !FileManager.default.fileExists(atPath: dir.path) {
            try FileManager.default.createDirectory(
                at: dir,
                withIntermediateDirectories: true,
                attributes: nil
            )
        }
    }
}

File: android/app/src/main/java/com/macecho/app/storage/TrustStore.kt
package com.macecho.app.storage

// Phase 9 — Local Trust Database (Encrypted Local Storage)
//
// TrustStore provides encrypted persistence for trusted device metadata per
// 04_SECURITY_MODEL.md §Local Trust Database and Architecture Decision 18.
//
// This is NOT the Android Keystore (that is KeystoreManager.kt).
// The Security Model explicitly separates:
//   - Private keys          → Android Keystore (KeystoreManager.kt)
//   - Trust metadata        → Encrypted local storage (this file)
//
// Trust metadata stored per entry (04_SECURITY_MODEL.md §Local Trust Database):
//   • Trusted Device Identifier
//   • Trusted Public Key (X25519 + Ed25519)
//   • Pairing Timestamp
//   • Trust Status
//   • Device Metadata (name, type)
//
// Encryption: Trust data is serialized as Java Properties, encrypted with a
// dedicated AES-256-GCM wrapping key stored in Android Keystore, and written
// to internal storage as a binary envelope.
//
// Must NOT contain:
//   - Pairing logic          → Phase 12 (TrustStore is populated by Phase 12)
//   - Authentication logic   → Phase 13
//   - Transport              → Phase 7
//   - Private key material   → KeystoreManager.kt
//   - Business logic of any kind
//
// Starts empty. Phase 12 (pairing) will call addOrUpdate() to populate it.
//
// File stored in: context.filesDir / "macecho_trust.enc"
// Format: [12-byte GCM nonce][4-byte ciphertext length][ciphertext]
//         where ciphertext decrypts to a UTF-8 Java Properties document

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import java.io.StringReader
import java.io.StringWriter
import java.security.KeyStore
import java.security.SecureRandom
import java.util.Base64
import java.util.Properties
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

// ---------------------------------------------------------------------------
// TrustStatus
// ---------------------------------------------------------------------------

/**
 * Represents the current trust status of a paired device.
 *
 * [TRUSTED] — The device has completed pairing and is recognized as a trusted peer.
 * [REVOKED] — Trust has been explicitly removed (unpairing, key change, or manual revocation).
 */
enum class TrustStatus { TRUSTED, REVOKED }

// ---------------------------------------------------------------------------
// TrustEntry — one record in the local trust database
// ---------------------------------------------------------------------------

/**
 * A single entry in the local trust database.
 *
 * Represents one trusted (or previously trusted) remote device. Populated
 * by Phase 12 (pairing) and validated by Phase 13 (authentication).
 *
 * Per 04_SECURITY_MODEL.md §Local Trust Database:
 *   - Never stores the other device's private key.
 *   - The backend has no authority to modify this record.
 *   - Only the local application may update trust records after successful
 *     security validation.
 *
 * @property trustedDeviceId         The peer's cryptographically random device ID.
 * @property trustedX25519PublicKeyBytes  Peer's X.509-encoded X25519 public key.
 * @property trustedEd25519PublicKeyBytes Peer's X.509-encoded Ed25519 public key.
 * @property pairingTimestampMs      Unix epoch ms when pairing completed.
 * @property trustStatus             Current trust status.
 * @property deviceName              Human-readable peer device name.
 * @property deviceType              Peer platform string (e.g. "MACOS").
 */
data class TrustEntry(
    val trustedDeviceId: String,
    val trustedX25519PublicKeyBytes: ByteArray,
    val trustedEd25519PublicKeyBytes: ByteArray,
    val pairingTimestampMs: Long,
    val trustStatus: TrustStatus,
    val deviceName: String,
    val deviceType: String,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TrustEntry) return false
        return trustedDeviceId == other.trustedDeviceId &&
            trustedX25519PublicKeyBytes.contentEquals(other.trustedX25519PublicKeyBytes) &&
            trustedEd25519PublicKeyBytes.contentEquals(other.trustedEd25519PublicKeyBytes) &&
            pairingTimestampMs == other.pairingTimestampMs &&
            trustStatus == other.trustStatus &&
            deviceName == other.deviceName &&
            deviceType == other.deviceType
    }

    override fun hashCode(): Int {
        var h = trustedDeviceId.hashCode()
        h = 31 * h + trustedX25519PublicKeyBytes.contentHashCode()
        h = 31 * h + trustedEd25519PublicKeyBytes.contentHashCode()
        h = 31 * h + pairingTimestampMs.hashCode()
        h = 31 * h + trustStatus.hashCode()
        h = 31 * h + deviceName.hashCode()
        h = 31 * h + deviceType.hashCode()
        return h
    }
}

// ---------------------------------------------------------------------------
// TrustStore
// ---------------------------------------------------------------------------

/**
 * Encrypted local storage for trusted device metadata.
 *
 * Architecture Decision 18: "Android stores trusted device metadata in
 * encrypted local storage." This is a separate store from the Android
 * Keystore used for private keys.
 *
 * Operations are synchronous. In production, callers should dispatch to a
 * background thread (e.g. Dispatchers.IO). Threading is not the concern of
 * this storage layer.
 *
 * @param context Android [Context] for internal file access.
 */
class TrustStore(private val context: Context) {

    private companion object {
        const val ANDROID_KEYSTORE_PROVIDER = "AndroidKeyStore"
        const val TRUST_KEY_ALIAS = "com.macecho.trust.wrapping"
        const val TRUST_FILE_NAME = "macecho_trust.enc"
        const val GCM_NONCE_BYTES = 12
        const val GCM_TAG_BITS = 128

        // Properties keys
        const val PROP_IDS = "entry.ids"
        const val PROP_X25519 = "x25519"
        const val PROP_ED25519 = "ed25519"
        const val PROP_TIMESTAMP = "timestamp"
        const val PROP_STATUS = "status"
        const val PROP_NAME = "name"
        const val PROP_TYPE = "type"
    }

    // CSPRNG — sole source of all random material (nonce generation)
    private val secureRandom = SecureRandom()

    // -----------------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------------

    /**
     * Adds or updates a trust entry.
     *
     * If an entry with [TrustEntry.trustedDeviceId] already exists, it is
     * replaced. Called by Phase 12 (pairing) after a successful pairing flow.
     */
    fun addOrUpdate(entry: TrustEntry) {
        val entries = loadEntries().toMutableMap()
        entries[entry.trustedDeviceId] = entry
        saveEntries(entries)
    }

    /**
     * Returns the trust entry for [deviceId], or null if not found.
     *
     * Called by Phase 13 (authentication) to verify an incoming connection.
     */
    fun get(deviceId: String): TrustEntry? = loadEntries()[deviceId]

    /**
     * Returns all trust entries in the store.
     *
     * Order is not guaranteed. The list may be empty if no devices have
     * been paired or all entries have been removed.
     */
    fun getAll(): List<TrustEntry> = loadEntries().values.toList()

    /**
     * Returns `true` if a trust entry exists for [deviceId].
     */
    fun contains(deviceId: String): Boolean = loadEntries().containsKey(deviceId)

    /**
     * Removes the trust entry for [deviceId].
     *
     * Safe to call when [deviceId] is not present (idempotent).
     * Called by Phase 12 (unpairing flow).
     */
    fun remove(deviceId: String) {
        val entries = loadEntries().toMutableMap()
        if (entries.remove(deviceId) != null) {
            saveEntries(entries)
        }
    }

    /**
     * Removes all trust entries. The trust store file is overwritten with
     * an empty encrypted store (not deleted).
     *
     * Used by application reset flows.
     */
    fun clear() {
        saveEntries(emptyMap())
    }

    /**
     * Returns the number of trust entries currently stored.
     */
    fun count(): Int = loadEntries().size

    // -----------------------------------------------------------------------
    // Internal serialization helpers (internal visibility for unit tests)
    // -----------------------------------------------------------------------

    /** Serializes [entries] to a Java Properties document (UTF-8 string bytes). */
    internal fun buildPropertiesBytes(entries: Map<String, TrustEntry>): ByteArray {
        val props = Properties()
        props[PROP_IDS] = entries.keys.joinToString(",")
        for ((id, entry) in entries) {
            fun key(field: String) = "entry.$id.$field"
            props[key(PROP_X25519)] = Base64.getEncoder().encodeToString(entry.trustedX25519PublicKeyBytes)
            props[key(PROP_ED25519)] = Base64.getEncoder().encodeToString(entry.trustedEd25519PublicKeyBytes)
            props[key(PROP_TIMESTAMP)] = entry.pairingTimestampMs.toString()
            props[key(PROP_STATUS)] = entry.trustStatus.name
            props[key(PROP_NAME)] = entry.deviceName
            props[key(PROP_TYPE)] = entry.deviceType
        }
        val sw = StringWriter()
        props.store(sw, null /* no header comment */)
        return sw.toString().toByteArray(Charsets.UTF_8)
    }

    /** Parses a Java Properties document back into trust entries. */
    internal fun parsePropertiesBytes(propsBytes: ByteArray): Map<String, TrustEntry> {
        val props = Properties().apply {
            load(StringReader(String(propsBytes, Charsets.UTF_8)))
        }
        val idsStr = props.getProperty(PROP_IDS) ?: return emptyMap()
        if (idsStr.isBlank()) return emptyMap()
        val ids = idsStr.split(",").map { it.trim() }.filter { it.isNotEmpty() }
        return ids.associateWith { id ->
            fun prop(field: String) = props.getProperty("entry.$id.$field")
                ?: error("Missing property entry.$id.$field")
            TrustEntry(
                trustedDeviceId = id,
                trustedX25519PublicKeyBytes = Base64.getDecoder().decode(prop(PROP_X25519)),
                trustedEd25519PublicKeyBytes = Base64.getDecoder().decode(prop(PROP_ED25519)),
                pairingTimestampMs = prop(PROP_TIMESTAMP).toLong(),
                trustStatus = TrustStatus.valueOf(prop(PROP_STATUS)),
                deviceName = prop(PROP_NAME),
                deviceType = prop(PROP_TYPE),
            )
        }
    }

    // -----------------------------------------------------------------------
    // Private implementation
    // -----------------------------------------------------------------------

    private fun trustFile(): File = File(context.filesDir, TRUST_FILE_NAME)

    private fun loadEntries(): Map<String, TrustEntry> {
        val file = trustFile()
        if (!file.exists()) return emptyMap()
        return runCatching {
            val key = getOrCreateTrustKey()
            val plaintext = decryptEnvelope(file.readBytes(), key)
            parsePropertiesBytes(plaintext)
        }.getOrDefault(emptyMap())
    }

    private fun saveEntries(entries: Map<String, TrustEntry>) {
        val key = getOrCreateTrustKey()
        val plaintext = buildPropertiesBytes(entries)
        val encrypted = encryptEnvelope(plaintext, key)
        trustFile().writeBytes(encrypted)
    }

    /**
     * Returns the existing AES trust wrapping key from Android Keystore, or
     * creates a new AES-256-GCM key in the Keystore (hardware-backed where available).
     *
     * This key is separate from the identity wrapping key in [KeystoreManager].
     * Per Architecture Decision 18: "Separating metadata into encrypted local
     * storage was preferred, since metadata does not require hardware-backed
     * key storage and benefits from simpler access patterns."
     */
    private fun getOrCreateTrustKey(): SecretKey {
        val ks = KeyStore.getInstance(ANDROID_KEYSTORE_PROVIDER).apply { load(null) }
        if (ks.containsAlias(TRUST_KEY_ALIAS)) {
            return (ks.getEntry(TRUST_KEY_ALIAS, null) as KeyStore.SecretKeyEntry).secretKey
        }
        val kg = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, ANDROID_KEYSTORE_PROVIDER)
        kg.init(
            KeyGenParameterSpec.Builder(
                TRUST_KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT,
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build(),
        )
        return kg.generateKey()
    }

    // Encrypted envelope format: [12-byte nonce][4-byte ciphertext length][ciphertext]
    private fun encryptEnvelope(plaintext: ByteArray, key: SecretKey): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key)
        val nonce = cipher.iv
        val ciphertext = cipher.doFinal(plaintext)
        val baos = ByteArrayOutputStream()
        DataOutputStream(baos).use { dos ->
            dos.write(nonce)
            dos.writeInt(ciphertext.size)
            dos.write(ciphertext)
        }
        return baos.toByteArray()
    }

    private fun decryptEnvelope(data: ByteArray, key: SecretKey): ByteArray {
        DataInputStream(ByteArrayInputStream(data)).use { dis ->
            val nonce = ByteArray(GCM_NONCE_BYTES).also { dis.readFully(it) }
            val ciphertext = ByteArray(dis.readInt()).also { dis.readFully(it) }
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(GCM_TAG_BITS, nonce))
            return cipher.doFinal(ciphertext)
        }
    }
}
